<a href="https://colab.research.google.com/github/ancestor9/2026_Fall_Learning-Langchain-AI-Agent/blob/main/CH1_LLM_Fundamentals_LangChain.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

In [1]:
# @title Getting Set Up with LangChain
import os

from google.colab import userdata
YOUR_API_KEY = userdata.get('Gemini-API-Key')

os.environ["GOOGLE_API_KEY"] = YOUR_API_KEY

In [2]:
import google.generativeai as genai

print("사용 가능한 Gemini 모델 목록:")
for m in genai.list_models():
    # 'generateContent' 메소드를 지원하는 모델만 필터링합니다.
    if 'generateContent' in m.supported_generation_methods:
        print(f"  Model Name: {m.name}, Supported Methods: {m.supported_generation_methods}")

사용 가능한 Gemini 모델 목록:
  Model Name: models/gemini-2.5-flash, Supported Methods: ['generateContent', 'countTokens', 'createCachedContent', 'batchGenerateContent']
  Model Name: models/gemini-2.5-pro, Supported Methods: ['generateContent', 'countTokens', 'createCachedContent', 'batchGenerateContent']
  Model Name: models/gemini-2.5-flash-preview-tts, Supported Methods: ['countTokens', 'generateContent']
  Model Name: models/gemini-2.5-pro-preview-tts, Supported Methods: ['countTokens', 'generateContent', 'batchGenerateContent']
  Model Name: models/gemma-4-26b-a4b-it, Supported Methods: ['generateContent', 'countTokens']
  Model Name: models/gemma-4-31b-it, Supported Methods: ['generateContent', 'countTokens']
  Model Name: models/gemini-flash-latest, Supported Methods: ['generateContent', 'countTokens', 'createCachedContent', 'batchGenerateContent']
  Model Name: models/gemini-flash-lite-latest, Supported Methods: ['generateContent', 'countTokens', 'createCachedContent', 'batchGenerateCo

/usr/local/lib/python3.12/dist-packages/google/colab/_import_hooks/_hook_injector.py:55: FutureWarning: 

All support for the `google.generativeai` package has ended. It will no longer be receiving 
updates or bug fixes. Please switch to the `google.genai` package as soon as possible.
See README for more details:

https://github.com/google-gemini/deprecated-generative-ai-python/blob/main/README.md

  loader.exec_module(module)


In [3]:
!pip install langchain-google-genai -q

   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 74.2/74.2 kB 1.0 MB/s eta 0:00:00
   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 570.0/570.0 kB 4.7 MB/s eta 0:00:00


In [4]:
from langchain_google_genai import ChatGoogleGenerativeAI

# 1. Gemini 모델 설정 (원하는 모델명을 기입합니다)
# 기본적으로 환경 변수에 설정된 GOOGLE_API_KEY를 자동으로 인식합니다.
model = ChatGoogleGenerativeAI(model="gemini-3.5-flash")

# 2. 인터페이스와 메서드는 OpenAI 때와 완벽히 동일합니다.
response = model.invoke("The sky is")
print(response.content)

[{'type': 'text', 'text': 'The sky can be described in a few different ways depending on what you have in mind:\n\n*   **Blue** during a clear day, which happens because the Earth\'s atmosphere scatters shorter wavelengths of light (blue and violet) from the sun more than other colors (a phenomenon called Rayleigh scattering).\n*   **"The limit"** (as the saying goes), meaning there are no bounds to what you can achieve.\n*   **A canvas of reds, pinks, and oranges** during sunrise and sunset.\n*   **Dark and starry** at night, revealing the vastness of space. \n\nHow does the sky look where you are right now?', 'extras': {'signature': 'EuQMCuEMARFNMg9OM41c+HMIJsT1fW+c2blu6wTM8Cb4FFre6W024ThTPeT6oyjKw9L1ybpOJMK4O45C90sNz9RGsx6kp3Y+USFl+8HEws+6MZRYP2qmS5/hAHIbEKiF0n4pjp3I2QhXdQwoJAukJhHhR7e7/+vVbAyoQPNMD3ZQPD/qPXAKIuU9xIu9ugQkM6B6Tlrbdz2xQCBLRtGsajgC2lP3k84d3nNzcKWfjoTrlfsM5ymrqHI4ZUTt7llmzijj0LU8/QCzEkWwjnSXVY2hdSN2owfrLnxV4jLegNhRcXZhTJ4U17o6jSat6MlEEidFuhXf6wFS5rxRMcetuDN76tJ4N6LiVFg1

In [5]:
print(response.text)

The sky can be described in a few different ways depending on what you have in mind:

*   **Blue** during a clear day, which happens because the Earth's atmosphere scatters shorter wavelengths of light (blue and violet) from the sun more than other colors (a phenomenon called Rayleigh scattering).
*   **"The limit"** (as the saying goes), meaning there are no bounds to what you can achieve.
*   **A canvas of reds, pinks, and oranges** during sunrise and sunset.
*   **Dark and starry** at night, revealing the vastness of space. 

How does the sky look where you are right now?


- 시스템 역할(System role) : Used for instructions the model should use to answer a user question(모델이 사용자 질문에 답변할 때 참조할 지침, 명령어에 사용)
- 사용자 역할 (User role): Used for the user’s query and any other content produced by the user(사용자의 질문 및 사용자가 작성한 모든 콘텐츠에 사용)
- 어시스턴트 역할 (Assistant role): Used for content generated by the model(모델이 생성한 답변 콘텐츠에 사용)

In [6]:
from langchain_core.messages import HumanMessage

prompt = [HumanMessage("What is the capital of France?")]

response = model.invoke(prompt)
print(response.content)
print(response.text)
print(response.json)

[{'type': 'text', 'text': 'The capital of France is **Paris**.', 'extras': {'signature': 'EvkCCvYCARFNMg8WZn8xT3Go97mCmaYtngjYQ/w7TEg4J6P7x7Vc5y72mRxzuJjCc9/zmWjqawGm8TqXfYZw3wS7bKK5GpTu/J/ZVPNeREmfbUMCH1dumjor+sW+OZ3/GC/u/t1U7t5F5wRZkxnRtbOHWTCZev/XipRHgnTC1WPW2zOyye0MjJEnGQvhOh9vsIvEp+A2AC8zhoZBOMgC9xieEKKbFdRPdixRIGXNyL9O3qoEZidLa0WHFjk/V7SEeQWBr+AD5LU08YTAwQQ0Him4+0NopqyTYNcQP8nZZXHosZup48m4cBSs8GXSuA3zHnpnfSVX8cN1U+cS6d3zK5HAukK6hYx4oupnrv32rBtwLn7zS/J7/oZKLLde6V/NJ5bydH9862kgWi7JdSB1ePFwvrc27Mpv7lvkiWYwPuV4c7XEfhDUqFnS7Nq5CdBj6xhdAFtr3sjlCNyO25kyQup+DjCddcKfLlOTXG0BT+AZuiCb8PXM8szTaBY='}}]
The capital of France is **Paris**.
<bound method BaseModel.json of AIMessage(content=[{'type': 'text', 'text': 'The capital of France is **Paris**.', 'extras': {'signature': 'EvkCCvYCARFNMg8WZn8xT3Go97mCmaYtngjYQ/w7TEg4J6P7x7Vc5y72mRxzuJjCc9/zmWjqawGm8TqXfYZw3wS7bKK5GpTu/J/ZVPNeREmfbUMCH1dumjor+sW+OZ3/GC/u/t1U7t5F5wRZkxnRtbOHWTCZev/XipRHgnTC1WPW2zOyye0MjJEnGQvhOh9vsIvEp+A2AC8zhoZBOMgC9xieEKKbF

In [7]:
print(response.json())

{"content":[{"type":"text","text":"The capital of France is **Paris**.","extras":{"signature":"EvkCCvYCARFNMg8WZn8xT3Go97mCmaYtngjYQ/w7TEg4J6P7x7Vc5y72mRxzuJjCc9/zmWjqawGm8TqXfYZw3wS7bKK5GpTu/J/ZVPNeREmfbUMCH1dumjor+sW+OZ3/GC/u/t1U7t5F5wRZkxnRtbOHWTCZev/XipRHgnTC1WPW2zOyye0MjJEnGQvhOh9vsIvEp+A2AC8zhoZBOMgC9xieEKKbFdRPdixRIGXNyL9O3qoEZidLa0WHFjk/V7SEeQWBr+AD5LU08YTAwQQ0Him4+0NopqyTYNcQP8nZZXHosZup48m4cBSs8GXSuA3zHnpnfSVX8cN1U+cS6d3zK5HAukK6hYx4oupnrv32rBtwLn7zS/J7/oZKLLde6V/NJ5bydH9862kgWi7JdSB1ePFwvrc27Mpv7lvkiWYwPuV4c7XEfhDUqFnS7Nq5CdBj6xhdAFtr3sjlCNyO25kyQup+DjCddcKfLlOTXG0BT+AZuiCb8PXM8szTaBY="}}],"additional_kwargs":{},"response_metadata":{"finish_reason":"STOP","model_name":"gemini-3.5-flash","safety_ratings":[],"model_provider":"google_genai"},"type":"ai","name":null,"id":"lc_run--01a020de-7f85-7521-8afa-7db41b3a9c78-0","tool_calls":[],"invalid_tool_calls":[],"usage_metadata":{"input_tokens":8,"output_tokens":72,"total_tokens":80,"input_token_details":{"cache_read":0},"output_tok

/tmp/ipykernel_1623/759109922.py:1: PydanticDeprecatedSince20: The `json` method is deprecated; use `model_dump_json` instead. Deprecated in Pydantic V2.0 to be removed in V3.0. See Pydantic V2 Migration Guide at https://errors.pydantic.dev/2.13/migration/
  print(response.json())


In [8]:
print(response.model_dump_json())

{"content":[{"type":"text","text":"The capital of France is **Paris**.","extras":{"signature":"EvkCCvYCARFNMg8WZn8xT3Go97mCmaYtngjYQ/w7TEg4J6P7x7Vc5y72mRxzuJjCc9/zmWjqawGm8TqXfYZw3wS7bKK5GpTu/J/ZVPNeREmfbUMCH1dumjor+sW+OZ3/GC/u/t1U7t5F5wRZkxnRtbOHWTCZev/XipRHgnTC1WPW2zOyye0MjJEnGQvhOh9vsIvEp+A2AC8zhoZBOMgC9xieEKKbFdRPdixRIGXNyL9O3qoEZidLa0WHFjk/V7SEeQWBr+AD5LU08YTAwQQ0Him4+0NopqyTYNcQP8nZZXHosZup48m4cBSs8GXSuA3zHnpnfSVX8cN1U+cS6d3zK5HAukK6hYx4oupnrv32rBtwLn7zS/J7/oZKLLde6V/NJ5bydH9862kgWi7JdSB1ePFwvrc27Mpv7lvkiWYwPuV4c7XEfhDUqFnS7Nq5CdBj6xhdAFtr3sjlCNyO25kyQup+DjCddcKfLlOTXG0BT+AZuiCb8PXM8szTaBY="}}],"additional_kwargs":{},"response_metadata":{"finish_reason":"STOP","model_name":"gemini-3.5-flash","safety_ratings":[],"model_provider":"google_genai"},"type":"ai","name":null,"id":"lc_run--01a020de-7f85-7521-8afa-7db41b3a9c78-0","tool_calls":[],"invalid_tool_calls":[],"usage_metadata":{"input_tokens":8,"output_tokens":72,"total_tokens":80,"input_token_details":{"cache_read":0},"output_tok

In [9]:
print(response.model_dump().keys())

dict_keys(['content', 'additional_kwargs', 'response_metadata', 'type', 'name', 'id', 'tool_calls', 'invalid_tool_calls', 'usage_metadata'])


In [10]:
print(response.model_dump()['tool_calls'])

[]


In [11]:
model

ChatGoogleGenerativeAI(metadata={'lc_versions': {'langchain-core': '1.6.0', 'langchain': '1.3.15', 'langchain-google-genai': '4.3.5'}}, profile={'name': 'Gemini 3.5 Flash', 'release_date': '2026-05-19', 'last_updated': '2026-05-19', 'open_weights': False, 'max_input_tokens': 1048576, 'max_output_tokens': 65536, 'text_inputs': True, 'image_inputs': True, 'audio_inputs': True, 'pdf_inputs': True, 'video_inputs': True, 'text_outputs': True, 'image_outputs': False, 'audio_outputs': False, 'video_outputs': False, 'reasoning_output': True, 'tool_calling': True, 'structured_output': True, 'attachment': True, 'temperature': True, 'image_url_inputs': True, 'image_tool_message': True, 'tool_choice': True, 'reasoning_effort_levels': ['minimal', 'low', 'medium', 'high'], 'reasoning_effort_default': 'medium'}, google_api_key=SecretStr('**********'), model='gemini-3.5-flash', temperature=None, client=<google.genai.client.Client object at 0x7ca19783f0b0>, default_metadata=(), model_kwargs={})

In [12]:
# @title  import LangChain’s OpenAI LLM wrapper to invoke a model prediction using a simple prompt:
from langchain_core.messages import SystemMessage

system_msg = SystemMessage(
    "You are a helpful assistant that responds to questions with three exclamation marks."
)
human_msg = HumanMessage("What is the capital of France?")

response = model.invoke([system_msg, human_msg])

print(response.text)

The capital of France is Paris!!!


In [19]:
from langchain_core.prompts import PromptTemplate

template = PromptTemplate.from_template("""Answer the question based on the context below.
If the question cannot be answered using the information provided, answer with "I don't know".

Context: {context}

Question: {question}

Answer: """)

response = template.invoke(
    {
        "context": '''The most recent advancements in NLP are being driven by Large Language Models (LLMs).
        These models outperform their smaller counterparts and have become invaluable for developers who are creating applications with NLP capabilities.
        Developers can tap into these models through Hugging Face's `transformers` library, or by utilizing OpenAI and
        Cohere's offerings through the `openai` and `cohere` libraries, respectively.''',

        "question": "Which model providers offer LLMs?",
    }
)

print(response)

text='Answer the question based on the context below. \nIf the question cannot be answered using the information provided, answer with "I don\'t know".\n\nContext: The most recent advancements in NLP are being driven by Large Language Models (LLMs). \n        These models outperform their smaller counterparts and have become invaluable for developers who are creating applications with NLP capabilities. \n        Developers can tap into these models through Hugging Face\'s `transformers` library, or by utilizing OpenAI and \n        Cohere\'s offerings through the `openai` and `cohere` libraries, respectively.\n\nQuestion: Which model providers offer LLMs?\n\nAnswer: '


In [14]:
print(response.text)

Answer the question based on the context below. 
If the question cannot be answered using the information provided, answer with "I don't know".

Context: The most recent advancements in NLP are being driven by Large Language Models (LLMs). These models outperform their smaller counterparts and have become invaluable for developers who are creating applications with NLP capabilities. Developers can tap into these models through Hugging Face's `transformers` library, or by utilizing OpenAI and Cohere's offerings through the `openai` and `cohere` libraries, respectively.

Question: Which model providers offer LLMs?

Answer: 


In [17]:
from langchain_core.prompts import PromptTemplate

# both `template` and `model` can be reused many times

template = PromptTemplate.from_template("""Answer the question based on the context below.
If the question cannot be answered using the information provided, answer with "I don't know".

Context: {context}

Question: {question}

Answer: """)

model = ChatGoogleGenerativeAI(model="gemini-3.5-flash")

# `prompt` and `completion` are the results of using template and model once

prompt = template.invoke(
    {
        "context": '''The most recent advancements in NLP are being driven by Large Language Models (LLMs).
        These models outperform their smaller counterparts and have become invaluable for developers who are creating applications with NLP capabilities.
        Developers can tap into these models through Hugging Face's `transformers` library, or by utilizing OpenAI and
        Cohere's offerings through the `openai` and `cohere` libraries, respectively.''',

        "question": "Which model providers offer LLMs?",
    }
)

response = model.invoke(prompt)
print(response)

content=[{'type': 'text', 'text': 'Based on the provided context, the model providers that offer LLMs are Hugging Face, OpenAI, and Cohere.', 'extras': {'signature': 'EoULCoILARFNMg9EAre2dtIGNl4Yit4OY4SnQtY0kRIBE3l2yC4vhmKZyaAbOsZsKe4FiPzsvQCZTmLXcMvcQ3xGFHyuxP28kHqe75u1c2KJNo6yeATDD68UngIoNCq4AbJ/dbVWRmPT1RWm9Y0bDZQl/DbdrloFo9zeQAyuIxfSQdEqOI3kb8FyUwsdObtc+bQz+cxijAdSIM/QOGB59fbq2cUp2QZLhCfMdLvFf+YQE5s4bwD2X5nBNqYRB/ijH+VoyNuYkM0dRa4Zh3XeSxWBERDONM+SEKGsFozdVDZcCaIe9q2B/wQ9JwhlmMcurvkH5PBgSBCkZORUqnC/FfRAeXW8mnZFiTrVGePsN4kDwpVObxy65BGhsANa1Km4ywwLvJnJf6LE9i5YEMmXAcjcUcuKJ17c/HwTd1/U561dL5mD15dNrJfpa8Jz85JpiV7XG8ZRo/9P4wsJv61YlWcjqc2Mg8UpUoTxGpbwx2lyWtaJFn/nQbHR8BduuWG8qN1j+b5TABjKROHj1uyvL+4RdyavRIE+PR7ZSVCxepzoLXx6gwg9CoMXMk3coWn2o1XAi9OIgmRM5V2UUXUjxuohW2VTsJK+FtinVvUiOboSd678bRCNTXCraQNNhlWg2gwAhSJP+FEBciYoVQ+LoANeYMkV+Rmr08ynFdMcZ8Rw3h7E1aSJRCPTSXFYRVwxUoPkQ1bQkiBEsKXM/U2BJeumuLwdqFLEpuP7UkkmIMRQrSRaIhh2SCBGyzy9AhYRUyY6Fu6ARP0QAzM8rGVfaufQskLzvERn70+IX86d4cvtlQMpJKTEPs3wpnny0d4uY

In [18]:
print(response.text)

Based on the provided context, the model providers that offer LLMs are Hugging Face, OpenAI, and Cohere.


In [21]:
# @title 🤖 LangChain Gemini Playground { run: "auto" }

# @markdown 우측 Form 영역에서 값을 변경하면 코드가 자동으로 반영됩니다.

prompt_input = "What is the capital of France?"  # @param {type:"string"}
model_name = "gemini-3.5-flash"  # @param ["gemini-2.5-flash", "gemini-3.5-flash"]
temperature_value = 0  # @param {type:"slider", min:0, max:1, step:0.1}


model = ChatGoogleGenerativeAI(
    model=model_name,
    temperature=temperature_value
)

response = model.invoke([HumanMessage(prompt_input)])
print("-" * 30)
print(response.text)

------------------------------
The capital of France is **Paris**.


### **[Groq Console](https://console.groq.com/home) 에 구글 아이디로 가입 후 API Key를 발급받고 ChatGroq 모듈을 호출한 후 코딩 실습 진행**

In [22]:
!pip install -qU langchain-groq -q

import os

# 발급받은 Groq API 키 입력
from google.colab import userdata
YOUR_API_KEY = userdata.get('groq')
os.environ["GROQ_API_KEY"] = YOUR_API_KEY

   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 137.5/137.5 kB 1.6 MB/s eta 0:00:00


In [24]:
from langchain_groq import ChatGroq
from langchain_core.messages import HumanMessage

# Llama 3 모델을 무료로 호출 (인터페이스는 완벽히 동일)
# model = ChatGroq(model="llama-3.1-8b-instant")
# 대체 모델로 교체
model = ChatGroq(model="openai/gpt-oss-20b")

response = model.invoke([HumanMessage("What is the capital of France?")])
print(response.text)

The capital of France is **Paris**.


In [25]:
from langchain_core.messages import SystemMessage, HumanMessage, AIMessage

# 시스템, 유저, 어시스턴트 역할을 배열로 묶어 대화 맥락을 주입
messages = [
    SystemMessage(content="당신은 유치원 선생님입니다. 모든 답변은 아이에게 말하듯 친절하게 하세요."),
    HumanMessage(content="달은 왜 밤에만 떠요?"),
    AIMessage(content="어두운 밤하늘을 우아하고 예쁘게 비춰주려고 밤에 은밀히 짜잔 하고 나타나는 거란다!"),
    HumanMessage(content="그럼 해님은요?")
]

response = model.invoke(messages)
print(response.content)

아, 해님은 언제나 빛나고 있는 거예요!  
지구는 바퀴처럼 돌면서, 그 바퀴가 돌 때마다 한쪽은 해님을 마주하고, 다른 쪽은 해님을 뒤로 하고 있죠.  

1. **지구가 돌고 있어요**  
   지구가 24시간에 한 바퀴를 돌면, 해님을 향해 있는 부분이 밝아지고, 해님을 향해 있지 않은 부분은 어두워져요. 그래서 해가 뜨고, 해가 지는 거예요.

2. **해는 항상 빛나고 있어요**  
   해는 하늘에 있는 별 중에서도 가장 밝은 별이랍니다. 밤이 되면 우리 눈이 해를 못 보게 되지만, 해는 여전히 빛나고 있어요. 그냥 지구가 바퀴를 돌면서 우리를 해와 멀리 떨어뜨리면 어둠이 오고, 다시 가까워지면 빛이 가득해지는 거예요.

3. **해가 떠오르면 낮이 시작돼요**  
   해가 떠오르면 하늘이 파랗게 빛나고, 동물들도 사람들도 밖에서 놀 수 있어요. 해가 지면 다시 밤이 되고, 달이 떠오르죠.

그래서 해님은 밤이 아니라 낮에 떠 있는 거예요! 해가 뜨면 우리에게 따뜻한 햇살을 주고, 해가 질 때는 하늘이 아름다운 색으로 물들어요. 🌞🌜


In [27]:
!pip install langchain-community -q

from langchain_community.chat_message_histories import ChatMessageHistory
from langchain_core.runnables.history import RunnableWithMessageHistory
from google.colab import userdata

# 1. API 키 및 모델 설정
os.environ["GROQ_API_KEY"] = userdata.get('groq')
model = ChatGroq(model="openai/gpt-oss-20b")

# 2. 대화 기록을 세션별로 저장할 딕셔너리와 관리 함수
'''
store는 세션 ID를 key로, 그 세션의 대화 기록(ChatMessageHistory 객체)을 value로 갖는 딕셔너리입니다.
get_session_history 함수는 특정 session_id에 해당하는 대화 기록을 꺼내오는 역할을 합니다.
처음 보는 session_id면 새 기록을 만들고, 이미 있으면 기존 기록을 반환합니다.
즉 사용자 A, 사용자 B가 각각 다른 session_id로 대화하면 서로의 대화 내용이 섞이지 않고 독립적으로 관리됩니다.
'''

store = {}

def get_session_history(session_id: str) -> ChatMessageHistory:
    if session_id not in store:
        store[session_id] = ChatMessageHistory()
    return store[session_id]

# 3. 랭체인의 래퍼(Wrapper)를 씌워 메모리 자동화 엔진 장착
'''
RunnableWithMessageHistory는 원래 모델(model)을 감싸서, 매번 대화 기록을 자동으로 주고받게 만들어주는 래퍼입니다.
원래는 대화를 이어가려면 이전 메시지들을 개발자가 직접 리스트에 모아서 매번 전달해야 하는데,
이 래퍼를 쓰면 get_session_history 함수를 통해 내부적으로 알아서 이전 대화를 불러오고, 새 응답을 저장까지 해줍니다.
'''

bot_with_history = RunnableWithMessageHistory(
    model,
    get_session_history,
)

# --- 🚀 실전 대화 테스트 (우리는 오직 질문만 던집니다) ---
'''
config에 session_id를 지정해서, 이 호출이 "user_123"이라는 대화방에서 이루어진다는 것을 알려줍니다.
내부적으로 일어나는 일:
get_session_history("user_123")으로 기존 기록(처음이라 비어있음)을 가져옴
기존 기록 + 새 HumanMessage를 합쳐서 모델에 전달
모델 응답을 받아서 그 기록에 자동으로 저장 (사용자 메시지 + AI 응답 둘 다)
'''

config = {"configurable": {"session_id": "user_123"}}

# 첫 번째 질문
response1 = bot_with_history.invoke([HumanMessage(content="내 이름은 상구야. 남양주에 살고 있어.")], config=config)
print("AI:", response1.content)

print("-" * 30)

# 두 번째 질문 (AI가 위에서 답변한 내용을 기억하고 동작함!)
'''
같은 session_id("user_123")로 다시 호출하면, 이번엔 저장소에 저장된 이전 대화(1번째 질문+답변)까지 함께 모델에 전달됩니다.
그래서 모델이 "이름은 상구, 사는 곳은 ○○"라고 이전 정보를 참조해서 답할 수 있는 겁니다.
'''

response2 = bot_with_history.invoke([HumanMessage(content="내 이름이 뭐라고 했지? 어디 산다고 했어?")], config=config)
print("AI:", response2.content)

/usr/local/lib/python3.12/dist-packages/IPython/core/interactiveshell.py:3553: LangChainDeprecationWarning: RunnableWithMessageHistory is deprecated. Use LangGraph's built-in persistence instead.
  exec(code_obj, self.user_global_ns, self.user_ns)


AI: 안녕하세요, 상구님! 남양주에서 살고 계시다니, 멋지네요. 혹시 궁금한 점이나 도움이 필요하신 것이 있으면 언제든 말씀해 주세요. 예를 들어, 지역 정보, 여행 팁, 혹은 다른 주제에 대해 물어보셔도 됩니다. 😊
------------------------------
AI: 네, 상구님!  
- **이름:** 상구  
- **거주지:** 남양주  

필요하신 다른 정보가 있으면 언제든 알려 주세요!


In [30]:
store.keys()

dict_keys(['user_123'])

In [34]:
store['user_123']

InMemoryChatMessageHistory(messages=[HumanMessage(content='내 이름은 상구야. 남양주에 살고 있어.', additional_kwargs={}, response_metadata={}), AIMessage(content='안녕하세요, 상구님! 남양주에서 살고 계시다니, 멋지네요. 혹시 궁금한 점이나 도움이 필요하신 것이 있으면 언제든 말씀해 주세요. 예를 들어, 지역 정보, 여행 팁, 혹은 다른 주제에 대해 물어보셔도 됩니다. 😊', additional_kwargs={'reasoning_content': 'The user says: "내 이름은 상구야. 남양주에 살고 있어." They are in Korean. They say "My name is Sanggu. I live in Namyangju." They haven\'t asked a question. We need to respond appropriately. The user might be introducing themselves. We can respond with a friendly greeting and maybe ask about their interests or how we can help. We should respond in Korean.'}, response_metadata={'token_usage': {'completion_tokens': 164, 'prompt_tokens': 86, 'total_tokens': 250, 'completion_time': 0.2303586, 'completion_tokens_details': {'reasoning_tokens': 86}, 'prompt_time': 0.004229439, 'prompt_tokens_details': None, 'queue_time': 0.156187601, 'total_time': 0.234588039}, 'model_name': 'openai/gpt-oss-20b', 'syst

In [35]:
store['user_123'].messages

[HumanMessage(content='내 이름은 상구야. 남양주에 살고 있어.', additional_kwargs={}, response_metadata={}),
 AIMessage(content='안녕하세요, 상구님! 남양주에서 살고 계시다니, 멋지네요. 혹시 궁금한 점이나 도움이 필요하신 것이 있으면 언제든 말씀해 주세요. 예를 들어, 지역 정보, 여행 팁, 혹은 다른 주제에 대해 물어보셔도 됩니다. 😊', additional_kwargs={'reasoning_content': 'The user says: "내 이름은 상구야. 남양주에 살고 있어." They are in Korean. They say "My name is Sanggu. I live in Namyangju." They haven\'t asked a question. We need to respond appropriately. The user might be introducing themselves. We can respond with a friendly greeting and maybe ask about their interests or how we can help. We should respond in Korean.'}, response_metadata={'token_usage': {'completion_tokens': 164, 'prompt_tokens': 86, 'total_tokens': 250, 'completion_time': 0.2303586, 'completion_tokens_details': {'reasoning_tokens': 86}, 'prompt_time': 0.004229439, 'prompt_tokens_details': None, 'queue_time': 0.156187601, 'total_time': 0.234588039}, 'model_name': 'openai/gpt-oss-20b', 'system_fingerprint': 'fp_a8c584dda7', '

- RunnableWithMessageHistory는 랭체인 최신 버전에서는 LangGraph의 체크포인터 방식으로 대체 권장되는 추세라, 새 프로젝트라면 langgraph의 MemorySaver 등을 쓰는 게 더 표준적인 방법일 수 있습니다.
- 지금 코드의 store는 파이썬 프로세스 안 메모리(dict)에만 저장되므로, 노트북을 재시작하거나 세션이 끊기면 대화 기록이 사라집니다. 실제 서비스라면 Redis, DB 등 영속 저장소로 바꿔야 합니다.

In [39]:
# @title Making LLM Prompts Reusable

from langchain_core.prompts import ChatPromptTemplate

template = ChatPromptTemplate.from_messages(
    [
        (
            "system",
            'Answer the question based on the context below. If the question cannot be answered using the information provided, answer with "I don\'t know".',
        ),
        ("human", "Context: {context}"),
        ("human", "Question: {question}"),
    ]
)

response = template.invoke(
    {
        "context": "The most recent advancements in NLP are being driven by Large Language Models (LLMs). These models outperform their smaller counterparts and have become invaluable for developers who are creating applications with NLP capabilities. Developers can tap into these models through Hugging Face's `transformers` library, or by utilizing OpenAI and Cohere's offerings through the `openai` and `cohere` libraries, respectively.",
        "question": "Which model providers offer LLMs?",
    }
)

print(response)

messages=[SystemMessage(content='Answer the question based on the context below. If the question cannot be answered using the information provided, answer with "I don\'t know".', additional_kwargs={}, response_metadata={}), HumanMessage(content="Context: The most recent advancements in NLP are being driven by Large Language Models (LLMs). These models outperform their smaller counterparts and have become invaluable for developers who are creating applications with NLP capabilities. Developers can tap into these models through Hugging Face's `transformers` library, or by utilizing OpenAI and Cohere's offerings through the `openai` and `cohere` libraries, respectively.", additional_kwargs={}, response_metadata={}), HumanMessage(content='Question: Which model providers offer LLMs?', additional_kwargs={}, response_metadata={})]


In [40]:
print(model.invoke(prompt))

content='Hugging\u202fFace, OpenAI, and Cohere.' additional_kwargs={'reasoning_content': 'We need to answer: Which model providers offer LLMs? Context: "Developers can tap into these models through Hugging Face\'s `transformers` library, or by utilizing OpenAI and Cohere\'s offerings through the `openai` and `cohere` libraries, respectively." So model providers: Hugging Face, OpenAI, Cohere. So answer: Hugging Face, OpenAI, Cohere.'} response_metadata={'token_usage': {'completion_tokens': 109, 'prompt_tokens': 199, 'total_tokens': 308, 'completion_time': 0.114383502, 'completion_tokens_details': {'reasoning_tokens': 87}, 'prompt_time': 0.011552351, 'prompt_tokens_details': None, 'queue_time': 0.311100848, 'total_time': 0.125935853}, 'model_name': 'openai/gpt-oss-20b', 'system_fingerprint': 'fp_639a5351c8', 'service_tier': 'on_demand', 'finish_reason': 'stop', 'logprobs': None, 'model_provider': 'groq'} id='lc_run--01a020f7-06fd-7bf3-bf74-6e62b24c1bad-0' tool_calls=[] invalid_tool_calls

In [41]:
print(model.invoke(prompt).content)

Hugging Face, OpenAI, and Cohere.


In [44]:
# @title Getting Specific Formats out of LLMs
from pydantic import BaseModel

class AnswerWithJustification(BaseModel):
    """An answer to the user's question along with justification for the answer."""

    answer: str
    """The answer to the user's question"""
    justification: str
    """Justification for the answer"""

llm = ChatGroq(model="openai/gpt-oss-20b", temperature=0)

structured_llm = llm.with_structured_output(AnswerWithJustification)

response = structured_llm.invoke(
    "What weighs more, a pound of bricks or a pound of feathers")

print(response)

answer='They weigh the same – both are a pound.' justification='A pound is a unit of mass, so a pound of bricks and a pound of feathers each have the same mass of one pound. The difference is in volume and density, not weight.'


In [45]:
# @title Other Machine-Readable Formats with Output Parsers

from langchain_core.output_parsers import CommaSeparatedListOutputParser

parser = CommaSeparatedListOutputParser()

response = parser.invoke("apple, banana, cherry")

print(response)

['apple', 'banana', 'cherry']


## **Assembling the Many Pieces of an LLM Application**
- Runnable = 하나의 작업(함수)이 객체화 된 것
> PromptTemplate, ChatGroq(ChatGoogleGenerativeAI, ChatOpenAI), OutputParser --> 모두 Runnable Intergace (method가 동일함, runnable.invoke() , runnable.batch(), runnable.stream())
- LCEL = Runnable들을 연결하는 파이프라인 문법
> chain = prompt | llm | parser --> Unix의 pipe와 동일


- 랭체인에서 "실행 가능한 단위"를 표준화한 인터페이스
- 모델, 프롬프트, 파서, 리트리버, 체인 등 거의 모든 랭체인 컴포넌트가 이 Runnable 인터페이스를 구현

- Runnable을 구현했다는 것은 다음 메서드들을 공통적으로 갖고 있다는 뜻

> - invoke() — 단일 입력에 대해 실행
> - batch() — 여러 입력을 한 번에 병렬 실행
> - stream() — 결과를 스트리밍으로 받기
> - ainvoke(), abatch(), astream() — 각각의 비동기(async) 버전

In [46]:
# @title Using the Runnable Interface
# @markdown - invoke() takes a single input and returns a single output.
# @markdown - batch() takes a list of outputs and returns a list of outputs.
# @markdown - stream() takes a single input and returns an iterator of parts of the output.

completion = model.invoke("Hi there!")

# Hi!
print(completion)

content='Hello! 👋 How can I help you today?' additional_kwargs={'reasoning_content': 'The user says "Hi there!" It\'s a greeting. We should respond politely. Probably ask how can help.'} response_metadata={'token_usage': {'completion_tokens': 43, 'prompt_tokens': 74, 'total_tokens': 117, 'completion_time': 0.047395316, 'completion_tokens_details': {'reasoning_tokens': 23}, 'prompt_time': 0.00830743, 'prompt_tokens_details': None, 'queue_time': 0.408437869, 'total_time': 0.055702746}, 'model_name': 'openai/gpt-oss-20b', 'system_fingerprint': 'fp_228717f27c', 'service_tier': 'on_demand', 'finish_reason': 'stop', 'logprobs': None, 'model_provider': 'groq'} id='lc_run--01a020fb-29fb-7d22-82ae-d62cdc91e559-0' tool_calls=[] invalid_tool_calls=[] usage_metadata={'input_tokens': 74, 'output_tokens': 43, 'total_tokens': 117, 'output_token_details': {'reasoning': 23}}


- completion = model.invoke("Hi there!")에 대한 LLM의 출력 구조와 내용에 대해 LLM에게 prompt해서 이해하자!

'''
content='Hello! 👋 How can I help you today?' additional_kwargs={'reasoning_content': 'The user says "Hi there!" It\'s a greeting. We should respond politely. Probably ask how can help.'} response_metadata={'token_usage': {'completion_tokens': 43, 'prompt_tokens': 74, 'total_tokens': 117, 'completion_time': 0.047395316, 'completion_tokens_details': {'reasoning_tokens': 23}, 'prompt_time': 0.00830743, 'prompt_tokens_details': None, 'queue_time': 0.408437869, 'total_time': 0.055702746}, 'model_name': 'openai/gpt-oss-20b', 'system_fingerprint': 'fp_228717f27c', 'service_tier': 'on_demand', 'finish_reason': 'stop', 'logprobs': None, 'model_provider': 'groq'} id='lc_run--01a020fb-29fb-7d22-82ae-d62cdc91e559-0' tool_calls=[] invalid_tool_calls=[] usage_metadata={'input_tokens': 74, 'output_tokens': 43, 'total_tokens': 117, 'output_token_details': {'reasoning': 23}}
'''
- content만 실제 답변이고 나머지는 전부 **부가 정보(메타데이터)** 입니다. 토큰 사용량 추적, 비용 계산, 디버깅, 응답 종료 사유 확인 등에 활용할 수 있습니다.


In [47]:
print(completion.content)

Hello! 👋 How can I help you today?


In [49]:
# @markdown - batch() takes a list of outputs and returns a list of outputs.

completions = model.batch(["Hi there!", "Bye!"])

# To print the content of each message:
for completion in completions:
    print(completion.content)

Hello! 👋 How can I help you today?
Goodbye! If you have any more questions in the future, feel free to reach out. Take care!


In [50]:
# @markdown - stream() takes a single input and returns an iterator of parts of the output.

for token in model.stream("Bye!"):
    print(token)

content='' additional_kwargs={} response_metadata={'model_provider': 'groq'} id='lc_run--01a02100-d3d9-7793-a9a1-df527b4107fe' tool_calls=[] invalid_tool_calls=[] tool_call_chunks=[]
content='' additional_kwargs={'reasoning_content': 'User'} response_metadata={'model_provider': 'groq'} id='lc_run--01a02100-d3d9-7793-a9a1-df527b4107fe' tool_calls=[] invalid_tool_calls=[] tool_call_chunks=[]
content='' additional_kwargs={'reasoning_content': ' says'} response_metadata={'model_provider': 'groq'} id='lc_run--01a02100-d3d9-7793-a9a1-df527b4107fe' tool_calls=[] invalid_tool_calls=[] tool_call_chunks=[]
content='' additional_kwargs={'reasoning_content': ' "'} response_metadata={'model_provider': 'groq'} id='lc_run--01a02100-d3d9-7793-a9a1-df527b4107fe' tool_calls=[] invalid_tool_calls=[] tool_call_chunks=[]
content='' additional_kwargs={'reasoning_content': 'Bye'} response_metadata={'model_provider': 'groq'} id='lc_run--01a02100-d3d9-7793-a9a1-df527b4107fe' tool_calls=[] invalid_tool_calls=[]

In [58]:
for chunk in model.stream("Bye!"):
    if chunk.content:
        print(chunk.content, end="", flush=True)

Goodbye! If you have any more questions in the future, feel free to reach out. Take care!

In [59]:
for chunk in model.stream("Bye!"):
    if chunk.content:
        print(chunk.content)
        print("*" * 50)

Good
**************************************************
bye
**************************************************
!
**************************************************
 If
**************************************************
 you
**************************************************
 have
**************************************************
 any
**************************************************
 more
**************************************************
 questions
**************************************************
 in
**************************************************
 the
**************************************************
 future
**************************************************
,
**************************************************
 feel
**************************************************
 free
**************************************************
 to
**************************************************
 reach
**************************************************
 out
**************************************

### Runnable들을 연결하는 2가지 방법
1.   Imperative: invoke()를 직접 하나씩 호출하며 실행 순서를 프로그래머가 명시
2.   Declarative(LCEL): prompt | llm | parser처럼 작업 흐름만 선언하면 LangChain이 실행 순서를 관리하고 필요한 기능(병렬 실행, 스트리밍, 비동기 등)을 일관된 방식으로 처리

> 두 방식의 결과는 같지만, Imperative는 "실행 방법을 직접 코딩"하는 방식(사용자정의 함수 실행)이고, LCEL은 "데이터 흐름을 선언"하는 방식이라는 점이 가장 큰 차이

- **Runnable Interface가 AsyncIO를 지원(LangChain은 OpenAI, Gemini, Anthropic, Ollama, Pinecone, Chroma, PostgreSQL 등 수많은 컴포넌트를 직렬 혹은 병렬로 연결해야 하기에 네트워크 요청(IO)은 비동기 방식을 지원)**

In [60]:
# @title Imperative Composition
# @markdown - Imperative composition is just a fancy name for writing the code you’re used to writing, composing these components into functions and classes. Here’s an example combining prompts, models, and output parsers:


from langchain_core.prompts import ChatPromptTemplate
from langchain_core.runnables import chain

# the building blocks

template = ChatPromptTemplate.from_messages(
    [
        ("system", "You are a helpful assistant."),
        ("human", "{question}"),
    ]
)

# combine them in a function
# @chain decorator adds the same Runnable interface for any function you write
# @chain 덕분에 chatbot 함수는 단순한 함수가 아니라 .
# invoke() 메서드를 사용할 수 있는 스마트한 LangChain 컴포넌트 역할

@chain
def chatbot(values):
    prompt = template.invoke(values)
    return model.invoke(prompt)

# use it
response = chatbot.invoke({"question": "Which model providers offer LLMs?"})
print(response.content)

Below is a quick‑reference list of the most prominent **model providers** that offer large‑language‑model (LLM) APIs or downloadable weights.  The list is grouped by the type of offering (cloud‑API, open‑source, or hybrid) and includes a short note on what each provider is best known for.

| # | Provider | Type | Typical Models / APIs | Notes |
|---|----------|------|-----------------------|-------|
| 1 | **OpenAI** | Cloud API | GPT‑4, GPT‑3.5‑turbo, GPT‑3.5‑turbo‑16k, Codex, DALL‑E, Whisper | The industry standard for general‑purpose LLMs; offers fine‑tuning and embeddings. |
| 2 | **Anthropic** | Cloud API | Claude 3, Claude 3.5, Claude 2 | Focus on safety‑first models; offers “Claude” family with different sizes. |
| 3 | **Microsoft Azure OpenAI** | Cloud API | GPT‑4, GPT‑3.5‑turbo, Codex, etc. | Same models as OpenAI but delivered via Azure’s infrastructure; includes Azure‑specific compliance and scaling. |
| 4 | **Google Cloud AI** | Cloud API | PaLM 2, Gemini (Beta), Vertex AI |

In [61]:
# @title Declarative Composition
# @markdown LCEL is a declarative language for composing LangChain components. LangChain compiles LCEL compositions to an optimized execution plan, with automatic paralleli‐zation, streaming, tracing, and async support.
template = ChatPromptTemplate.from_messages(
    [
        ("system", "You are a helpful assistant."),
        ("human", "{question}"),
    ]
)


@chain
def chatbot(values):
    prompt = template.invoke(values)
    for token in model.stream(prompt):
        yield token


for part in chatbot.stream({"question": "Which model providers offer LLMs?"}):
    print(part)

content='' additional_kwargs={} response_metadata={'model_provider': 'groq'} id='lc_run--01a02108-ec58-7b81-bc9d-a38b03affb05' tool_calls=[] invalid_tool_calls=[] tool_call_chunks=[]
content='' additional_kwargs={'reasoning_content': 'The'} response_metadata={'model_provider': 'groq'} id='lc_run--01a02108-ec58-7b81-bc9d-a38b03affb05' tool_calls=[] invalid_tool_calls=[] tool_call_chunks=[]
content='' additional_kwargs={'reasoning_content': ' user'} response_metadata={'model_provider': 'groq'} id='lc_run--01a02108-ec58-7b81-bc9d-a38b03affb05' tool_calls=[] invalid_tool_calls=[] tool_call_chunks=[]
content='' additional_kwargs={'reasoning_content': ' asks'} response_metadata={'model_provider': 'groq'} id='lc_run--01a02108-ec58-7b81-bc9d-a38b03affb05' tool_calls=[] invalid_tool_calls=[] tool_call_chunks=[]
content='' additional_kwargs={'reasoning_content': ':'} response_metadata={'model_provider': 'groq'} id='lc_run--01a02108-ec58-7b81-bc9d-a38b03affb05' tool_calls=[] invalid_tool_calls=[]

In [63]:
input_data = {"question": "Which model providers offer LLMs?"}
formatted_prompt = template.invoke(input_data)
for chunk in model.stream(formatted_prompt):
    if chunk.content:
        print(chunk.content, end="")

**Major providers that offer large‑language‑model (LLM) APIs or services**

| Provider | Typical LLMs / Models | Notes |
|----------|----------------------|-------|
| **OpenAI** | GPT‑4, GPT‑3.5‑turbo, GPT‑3.5‑turbo‑16k, GPT‑3.5‑turbo‑0613, etc. | Cloud‑based API, fine‑tuning (GPT‑4‑Turbo‑ft), embeddings, image‑generation (DALL‑E). |
| **Anthropic** | Claude 3, Claude 3.5, Claude 2, Claude 2.1 | API via Anthropic or integrated into Anthropic’s own platform. |
| **Microsoft Azure OpenAI** | GPT‑4, GPT‑3.5‑turbo, Codex, etc. | Same OpenAI models, but hosted on Azure with enterprise‑grade SLAs. |
| **Google Cloud** | PaLM 2 (text‑Bison, code‑Bison), Gemini (preview), Vertex AI | Google’s own LLMs, available via Vertex AI or Cloud API. |
| **Amazon Bedrock** | Claude, GPT‑4, Llama‑2, Mistral, etc. | Multi‑model platform; you can choose from several providers. |
| **Cohere** | Command R, Command R‑2, Command R‑3, etc. | Focus on retrieval‑augmented generation and embeddings. |
| **AI21 Labs

In [67]:
# @title **@chain 데코레이터를 이용해서 비동기(async) 커스텀 체인을 만들고 실행**

@chain
async def chatbot(values):
    prompt = await template.ainvoke(values)
    return await model.ainvoke(prompt)


async def main():
    return await chatbot.ainvoke({"question": "Which model providers offer LLMs?"})


# In Colab, we can directly await top-level async functions.
print(await main())

content='Below is a quick‑reference list of the most prominent **model providers** that offer large‑language‑model (LLM) APIs or downloadable weights.  The list is grouped by the type of offering (cloud‑API, open‑source, or hybrid) and includes a short note on what each provider is best known for.\n\n| # | Provider | Type of Offering | Notable Models | Typical Use‑Cases |\n|---|----------|------------------|----------------|-------------------|\n| 1 | **OpenAI** | Cloud API | GPT‑4, GPT‑3.5‑turbo, GPT‑3.5‑turbo‑16k | Chat, code generation, summarization, content creation |\n| 2 | **Anthropic** | Cloud API | Claude\u202f3, Claude\u202f3.5 Sonnet | Conversational agents, safety‑focused chat, policy‑compliant content |\n| 3 | **Microsoft Azure OpenAI** | Cloud API (OpenAI models on Azure) | GPT‑4, GPT‑3.5, Codex | Enterprise‑grade integration, compliance, Azure‑specific tooling |\n| 4 | **Google Cloud AI** | Cloud API | PaLM\u202f2, Gemini (Beta) | Search‑enhanced chat, code, multimodal t

In [65]:
# combine them with the | operator
chatbot = template | model

# use it
response = chatbot.invoke({"question": "Which model providers offer LLMs?"})
print(response.content)

Below is a quick‑reference list of the most prominent **model providers that offer large‑language‑model (LLM) APIs, services, or downloadable weights**.  The list is grouped by the type of offering (commercial API, cloud‑platform integration, or open‑source release) and is current as of August 2026.

| Category | Provider | Typical LLMs / Models | Notes |
|----------|----------|-----------------------|-------|
| **Commercial API (directly from the provider)** | **OpenAI** | GPT‑4, GPT‑4o, GPT‑3.5‑turbo, GPT‑4‑vision, etc. | Most widely used; fine‑tuning via “Fine‑Tuning” and “Embeddings” endpoints. |
| | **Anthropic** | Claude 3.5 Sonnet, Claude 3 Opus, Claude 3 Haiku, Claude 2.1 | Strong safety & alignment focus; “Claude” brand. |
| | **Microsoft Azure OpenAI** | GPT‑4, GPT‑4o, GPT‑3.5‑turbo, etc. | Same models as OpenAI but delivered via Azure; includes Azure‑specific compliance & scaling. |
| | **Google Cloud AI** | PaLM 2 (Chat‑PaLM, Code‑PaLM), Gemini 1.0, Gemini 1.5 | Google’s fl

In [66]:
# streaming (token 형태로 출력)
for part in chatbot.stream({"question": "Which model providers offer LLMs?"}):
    print(part)

content='' additional_kwargs={} response_metadata={'model_provider': 'groq'} id='lc_run--01a0210c-09d8-7461-a582-64a39e96d229' tool_calls=[] invalid_tool_calls=[] tool_call_chunks=[]
content='' additional_kwargs={'reasoning_content': 'The'} response_metadata={'model_provider': 'groq'} id='lc_run--01a0210c-09d8-7461-a582-64a39e96d229' tool_calls=[] invalid_tool_calls=[] tool_call_chunks=[]
content='' additional_kwargs={'reasoning_content': ' user'} response_metadata={'model_provider': 'groq'} id='lc_run--01a0210c-09d8-7461-a582-64a39e96d229' tool_calls=[] invalid_tool_calls=[] tool_call_chunks=[]
content='' additional_kwargs={'reasoning_content': ' asks'} response_metadata={'model_provider': 'groq'} id='lc_run--01a0210c-09d8-7461-a582-64a39e96d229' tool_calls=[] invalid_tool_calls=[] tool_call_chunks=[]
content='' additional_kwargs={'reasoning_content': ':'} response_metadata={'model_provider': 'groq'} id='lc_run--01a0210c-09d8-7461-a582-64a39e96d229' tool_calls=[] invalid_tool_calls=[]